### PR TITLE
Allow comparisons to refine Optional types without strict-optional

### LINKED ISSUE #4520: explicit `x is None` should narrow union candidates like `isinstance(x, type(None))`
```
from typing import Optional, List  # noqa


def union_test1(x):
    # type: (Optional[List[int]]) -> Optional[int]
    if x is None:
        return x
    else:
        return len(x)


def union_test2(x):
    # type: (Optional[List[int]]) -> Optional[int]
    if isinstance(x, type(None)):
        return x
    else:
        return len(x)
```

`union_test1` with default options gives the error `Incompatible return value type (got "Optional[List[int]]", expected "Optional[int]")` (with `--strict-optional`, it does not give this error). `union_test2` typechecks just fine. This strikes me as highly counterintuitive; even though (as the docs say) `None` is a valid value for each type, if one of the above checks allows the typechecker to learn that `x` is `None` and *therefore* not a list, it's hard to see why the other doesn't.

Failing that, I would suggest at least a documentation change to point out both the gotcha embodied in `union_test1` and the workaround in `union_test2`.

### PR DESCRIPTION
**Pull request by msullivan**
Fixes #4520.

### CODE REVIEW

**Comment by gvanrossum** [comment]
(Let's hold off on this a bit, we've discovered some problems with it in our own codebase.)

**Comment by ilevkivskyi** [comment]
> (Let's hold off on this a bit, we've discovered some problems with it in our own codebase.)

Please let me know when the problems are solved, so that I can proceed with a review.

**Comment by gvanrossum** [comment]
IIUC that's in @msullivan's court.

**Comment by msullivan** [comment]
Before this can land, we need to fix #3526, since this change exposes a bunch of instances of it.

**Comment by msullivan** [comment]
@ilevkivskyi this is ready for review now

File: mypy/binder.py [hunk 1]
```diff
@@ -271,11 +271,10 @@ def assign_type(self, expr: Expression,
                 and not restrict_any):
```
Comment (ilevkivskyi) [inline mypy/binder.py]: This doesn't need to be on a separate line.

**Comment by msullivan** [comment]
This causes a small handful of errors in internal codebases that will be easy to fix. A couple were weird second-order effects in which doing the refinement meant that a type was no longer a Union and so behaved differently when computing types in a ternary expression. Sigh.

**Pull request merged**